HidPkg/UefiHidDxe/src/pointer.rs

```diff
@@ -0,0 +1,513 @@
+  controller: efi::Handle,
+  driver_binding: &driver_binding::Protocol,
+) -> Result<*mut HidContext, efi::Status> {
+  // retrieve a reference to boot services.
+  // Safety: BOOT_SERVICES must have been initialized to point to the UEFI Boot Services table.
+  // Caller should have ensured this, so just expect on failure.
+  let boot_services = unsafe { BOOT_SERVICES.as_mut().expect("BOOT_SERVICES not properly initialized") };
+
+  let mut absolute_pointer_ptr: *mut absolute_pointer::protocol::Protocol = core::ptr::null_mut();
+  let status = (boot_services.open_protocol)(
+    controller,
+    &absolute_pointer::protocol::PROTOCOL_GUID as *const efi::Guid as *mut efi::Guid,
+    core::ptr::addr_of_mut!(absolute_pointer_ptr) as *mut *mut c_void,
+    driver_binding.driver_binding_handle,
+    controller,
+    system::OPEN_PROTOCOL_GET_PROTOCOL,
+  );
+
+  match status {
+    efi::Status::SUCCESS => {
+      // retrieve a reference to the pointer context.
+      // Safety: `absolute_pointer_ptr` must point to an instance of absolute_pointer that is contained in a
+      // PointerContext struct. The following is the equivalent of the `CR` (contained record) macro in the EDK2 C
+      // reference implementation.
+      let context_ptr = unsafe { (absolute_pointer_ptr as *mut u8).sub(offset_of!(PointerContext, absolute_pointer)) }
+        as *mut PointerContext;
+      return Ok(unsafe { (*context_ptr).hid_context });
+    }
+    err => return Err(err),
+  }
+}
+
+// event handler for wait_for_pointer event that is part of the absolute pointer interface.
+extern "efiapi" fn wait_for_pointer(event: efi::Event, context: *mut c_void) {
+  // retrieve a reference to boot services.
+  // Safety: BOOT_SERVICES must have been initialized to point to the UEFI Boot Services table.
+  // Caller should have ensured this, so just expect on failure.
+  let boot_services = unsafe { BOOT_SERVICES.as_mut().expect("BOOT_SERVICES not properly initialized") };
+
+  // retrieve a reference to pointer context.
+  // Safety: Pointer Context must have been initialized before this event could fire, so just expect on failure.
+  let pointer_context = unsafe { (context as *mut PointerContext).as_mut().expect("Pointer Context is bad.") };
+
+  if pointer_context.handler.state_changed {
+    (boot_services.signal_event)(event);
+  }
+}
+
+// resets the pointer state - part of the absolute pointer interface.
+extern "efiapi" fn absolute_pointer_reset(
+  this: *const absolute_pointer::protocol::Protocol,
+  _extended_verification: bool,
+) -> efi::Status {
+  // retrieve a reference to the pointer context.
+  // Safety: `this` must point to an instance of absolute_pointer that is contained in a PointerContext struct.
+  // the following is the equivalent of the `CR` (contained record) macro in the EDK2 reference implementation.
+  let context_ptr =
+    unsafe { (this as *mut u8).sub(offset_of!(PointerContext, absolute_pointer)) } as *mut PointerContext;

```

Comment:
Would it still make sense to explicitly check whether `this` is null before using it here?

Reply:
yep - added check here and in absolute_pointer_get_state